storage/storage_dest.go

```diff
@@ -365,13 +365,15 @@ func (s *storageImageDestination) tryReusingBlobAsPending(blobDigest digest.Dige
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	if options.SrcRef != nil {
+	if options.SrcRef != nil && options.TOCDigest != "" && options.TOCDigest.Validate() == nil {
 		// Check if we have the layer in the underlying additional layer store.
-		aLayer, err := s.imageRef.transport.store.LookupAdditionalLayer(blobDigest, options.SrcRef.String())
+		aLayer, err := s.imageRef.transport.store.LookupAdditionalLayer(options.TOCDigest, options.SrcRef.String())
 		if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
 			return false, private.ReusedBlob{}, fmt.Errorf(`looking for compressed layers with digest %q and labels: %w`, blobDigest, err)
 		} else if err == nil {
-			s.lockProtected.blobDiffIDs[blobDigest] = aLayer.UncompressedDigest()
+			if d := aLayer.TOCDigest(); d != "" && options.LayerIndex != nil {
+				s.lockProtected.indexToTOCDigest[*options.LayerIndex] = d
+			}
```

Comment:
- This _must_ define a layer ID as documented in `storageImageDestinationLockProtected`. If no suitable value is available (can that ever happen??), we just can’t report success here, and we must fall through to the other reuse / non-reuse paths.
- The layer ID  _must not_ be based on an untrusted value. If the only value the backend is using is the TOC digest, we must not use `blobDigest` for any purpose.
- So, `blobAdditionalLayer` should be indexed either by the TOC digest, or by by `LayerIndex`; and the code paths can only be reachable if that data is available.

Reply:
Thanks for the comment. Fixed.